pkg/statedb/reconciler/types.go

```diff
@@ -0,0 +1,207 @@
+	// the given value.
+	WithObjectStatus func(Obj, Status) Obj
+
+	// StatusIndex is the index for looking of objects by StatusKind. The index
+	// is created with NewStatusIndex.
+	StatusIndex statedb.Index[Obj, StatusKind]
+}
+
+func (cfg Config[Obj]) validate() error {
+	if cfg.GetObjectStatus == nil {
+		return fmt.Errorf("%T.GetObjectStatus cannot be nil", cfg)
+	}
+	if cfg.WithObjectStatus == nil {
+		return fmt.Errorf("%T.WithObjectStatus cannot be nil", cfg)
+	}
+	if cfg.StatusIndex.Name == "" {
+		return fmt.Errorf("%T.StatusIndex needs to be defined", cfg)
+	}
+	if cfg.IncrementalBatchSize <= 0 {
+		return fmt.Errorf("%T.IncrementalBatchSize needs to be >0", cfg)
+	}
+	if cfg.FullReconcilationInterval < 0 {
+		return fmt.Errorf("%T.FullReconcilationInterval must be >=0", cfg)
+	}
+	if cfg.RetryBackoffMaxDuration <= 0 {
+		return fmt.Errorf("%T.RetryBackoffMaxDuration must be >0", cfg)
+	}
+	if cfg.RetryBackoffMinDuration <= 0 {
+		return fmt.Errorf("%T.RetryBackoffMinDuration must be >0", cfg)
+	}
+	return nil
+}
+
+// Operations defines how to reconcile an object.
+//
+// Each operation is given a context that limits the lifetime of the operation
+// and a ReadTxn to allow for the option of looking up realized state from another
+// statedb table as an optimization (main use-case is reconciling routes against
+// Table[Route] to avoid a syscall per route).
+type Operations[Obj any] interface {
+	// Update the object in the target. If the operation is long-running it should
+	// abort if context is cancelled. Should return an error if the operation fails.
+	// The reconciler will retry the operation again at a later time, potentially
+	// with a new version of the object. The operation should thus be idempotent.
+	//
+	// Update is used both for incremental and full reconciliation. Incremental
+	// reconciliation is performed when the desired state is updated. A full
+	// reconciliation is done periodically by calling 'Update' on all objects.
+	//
+	// 'changed' must be true if the Update resulted in a changed target state.
+	// This allows detecting whether full reconciliation caught an out-of-sync
+	// target state and for tracking this in metrics.
+	Update(context.Context, statedb.ReadTxn, Obj) (changed bool, err error)
+
+	// Delete the object in the target. Same semantics as with Update.
+	// Deleting a non-existing object is not an error and returns nil.
+	Delete(context.Context, statedb.ReadTxn, Obj) error
+
+	// Prune undesired state. It is given an iterator for the full set of
+	// desired objects. The implementation should diff the desired state against
+	// the realized state to find things to prune.
+	// Invoked during full reconciliation before the individual objects are Update()'d.
+	//
+	// Unlike failed Update()'s a failed Prune() operation is not retried until
+	// the next full reconciliation round.
+	Prune(context.Context, statedb.ReadTxn, statedb.Iterator[Obj]) error
```

Comment:
What's the purpose of `statedb.ReadTxn` if we have `statedb.Iterator[Obj]`? For example the `Prune` implementation done in pkg/statedb/reconciler/example/ops.go, isn't using the `statedb.ReadTxn`.

Reply:
The only use-case thus far has been in the route reconciler (not part of this PR though). There the operations could check against `Table[*Route]` (a mirror of the realized state) instead of doing syscalls. I'll make sure this gets documented!

Reply:
Ah looks like I did document it as part of the doc comment of `Operations`